### PR TITLE
Performance tests for `DataCatalog`

### LINKED ISSUE #4125: [Stress Testing] - Data Catalog and Config Loader
## Description

To create stress test for individual components:
- DataCatalog
- ConfigLoader

> As for DataCatalog, the most important thing is to test it within the pipeline, CLI and separately by simulating scenarios when calling some methods as (add_feed_dict). Where the tests themselves should include different sets and combinations of parameters, datasets and patterns. https://github.com/kedro-org/kedro/issues/3957#issuecomment-2299105368

The tickets combined the two components as I believe the setup will be similar, or at least there need to be some mechanism to setup different datasets first so I thought it makes sense to bundle these two components together.

## Context

https://github.com/kedro-org/kedro/issues/3957#issuecomment-2298861806

> ## Component stress test:
> - The main goal for this is to benchmark performance of individual component, this will inform if refactoring work has positive/negative impact. Currently we only check if test pass, so we have no idea if a change may slow down performance. We have done this in the past but usually ad-hoc basis, we should run this regularly (or at least per release).
> 
> The direction of this is simple, we want to make measure the change of time against # number of entries. We would start with Datasets and Catalog, as this fits in the `DataCatalog2.0` work and will be immediately useful.
> 
> 
> - DataCatalog (test # of datasets with catalog.yml & dataset factory)
> - ConfigLoader(# of parameters)
> - Optional: pipelines generated in loops (Dynamic pipeline)
> 
> This can address:
> - https://github.com/kedro-org/kedro/issues/3893
> - DataCatalog 2.0
>

### PR DESCRIPTION
**Pull request by ankatiyar**
## Description

Partially resolve #4125 

## Development notes

Added performance tests for `DataCatalog`
I'm wondering if it's worth adding them for `KedroDataCatalog` as well, since the interface is still in active development?
Ignore `benchmark_ocl.py` since that is under review in #4225 

To run locally: `asv run` or `asv run --quick`

## Developer Certificate of Origin
We need all contributions to comply with the [Developer Certificate of Origin (DCO)](https://developercertificate.org/). All commits must be signed off by including a `Signed-off-by` line in the commit message. [See our wiki for guidance](https://github.com/kedro-org/kedro/wiki/Guidelines-for-contributing-developers/).

If your PR is blocked due to unsigned commits, then you must follow the instructions under "Rebase the branch" on the GitHub Checks page for your PR. This will retroactively add the sign-off to all unsigned commits and allow the DCO check to pass.

## Checklist

- [ ] Read the [contributing](https://github.com/kedro-org/kedro/blob/main/CONTRIBUTING.md) guidelines
- [ ] Signed off each commit with a [Developer Certificate of Origin (DCO)](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/managing-repository-settings/managing-the-commit-signoff-policy-for-your-repository)
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added a description of this change in the [`RELEASE.md`](https://github.com/kedro-org/kedro/blob/main/RELEASE.md) file
- [ ] Added tests to cover my changes
- [ ] Checked if this change will affect Kedro-Viz, and if so, communicated that with the Viz team

### CODE REVIEW

**Review by noklam — COMMENTED** [review]
add KedroDataCatalog makes sense, even if it's one-off I would be interested to see how it compares to the current one.

File: benchmarks/benchmark_ocl.py
Comment (ElenaKhaustova) [inline benchmarks/benchmark_ocl.py:36]: Is the indentation here correct?

File: benchmarks/benchmark_ocl.py
Comment (ElenaKhaustova) [inline benchmarks/benchmark_ocl.py:66]: Can we maybe replace this with a loop?

**Comment by ElenaKhaustova** [comment]
Could we wait until we merge https://github.com/kedro-org/kedro/pull/4218 for `KedroDataCatalog`? After that, the interface won't change much, so it might be easier. So I would bound this PR with the old catalog only, but up to the developer.

File: benchmarks/benchmark_ocl.py
Comment (ankatiyar) [reply benchmarks/benchmark_ocl.py]: will fix this in this PR

File: benchmarks/benchmark_ocl.py
Comment (ankatiyar) [reply benchmarks/benchmark_ocl.py]: yea let me see if i can simplify it

**Review by merelcht — APPROVED** [review]
⭐

**Pull request merged**
